### PR TITLE
[TFTRT] Add Dynamic Shape Testing and fix Explicit Batch Mode for ConvertUnpack

### PR DESCRIPTION
**Pull request by DEKHTIARJonathan**
Fixed ConvertSplitHelper to make ConvertUnpack work for explicit batch mode.
Add dynamic shape testing for ConvertUnpack.

@bixia1 @tfeher for review

Feature Tracker: #45481

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -5091,41 +5098,68 @@ Status ConvertSplitHelper(OpConverterParams* params,
...
   for (int i = 0; i < num_splits; ++i) {
-    begin[trt_axis + 1] = i * split_size_on_axis;
+    const int tf_axis = trt_axis + (params->use_implicit_batch ? 1 : 0);
+    begin[tf_axis] = i * split_size_on_axis;
     TF_RETURN_IF_ERROR(ConvertStridedSliceHelper(
-        params, input, begin, size, stride, final_shape_for_unpack_ptr, i));
+        params, input, begin, size, stride, final_shape_for_unpack_ptr,
+        /*op_instance=*/i));
+    if (! params->validation_only){
+    }
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:5149]: Not needed?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -5091,41 +5098,68 @@ Status ConvertSplitHelper(OpConverterParams* params,
...
+  }
+  if (params->validation_only) return Status::OK();
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:5151]: I should remove this for these reasons:
     The SqueezeTensor converter can handle validataion_only mode.
      The SqueezeTensor converter may return error, in that case, we should not convert and returning Status::OK() here premature and wrong.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 3]
```diff
@@ -6346,156 +6347,185 @@ auto get_unpack_nodedef = [](DataType dtype, int num, int axis) -> NodeDef {
...
+// TODO: Reactivate when INT32 Segfault fixed
 #if IS_TRT_VERSION_GE(5, 1, 3, 1)
-  TestConvertUnpack<DT_INT32>(this);
+TEST_P(OpConverter_FP32_FP16_INT32_Test, ConvertUnpack) {
+#else
+TEST_P(OpConverter_FP32_FP16_Test, ConvertUnpack) {
 #endif
+  // We need to skip error testing for Dynamic Shape mode, as it is impossible
+  // to convert Unpack in Dynamic Shape Mode.
+  // `Unimplemented: Strided slice op not implemented for dynamic shape input`
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:6396]: Sorry, I don't understand this line. What is the purpose of this line?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 1 above)
Comment (DEKHTIARJonathan) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:5149]: Indeed debug leftovers. Removing

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 3 above)
Comment (DEKHTIARJonathan) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:6396]: Error expected from the converter if you try dynamic shape inputs.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 2 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:5149]: `SqueezeTensor` seems to handle `validation_only` mode, but in practice we do not use it for validation in any other converter. I believe we cannot use it here either:
- The first arg (`params->outputs->at(i)`) is not created by `ConvertStridedSliceHelper` during validation.
- I am not sure if `params->converter` exists in `validation_mode`.

**Review by bixia1 — APPROVED** [review]

**Review by bixia1 — CHANGES_REQUESTED** [review]
The added test didn't pass:

third_party/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:129
No Diff
Inline Diff
Line Diff
Token Diff
Char Diff
Ignore White Space
Value of: status.error_message()
Expected: has substring "Dimension 3 of size 3 is not evenly divisble by 2, at my_split"
  Actual: "Dimension 3 of size 3 is not evenly divisible by 2, at my_split"
Invalid argument: Dimension 3 of size 3 is not evenly divisible by 2, at my_split
Stack trace:
0x55fb2b36f5ad: tensorflow::tensorrt::convert::ExpectStatus(tensorflow::Status, tensorflow::error::Code, char const*) @ ../sysdeps/x86_64/start.S:121
0x55fb2b385435: tensorflow::tensorrt::convert::OpConverterTest::RunValidationAndConversion(tensorflow::NodeDef const&, tensorflow::error::Code, char const*, bool) @ ??:??
0x55fb2b3da77e: tensorflow::tensorrt::convert::OpConverterTest_ConvertSplit_Test::TestBody() @ ??:??
0x55fb2c886521: testing::Test::Run() @ ??:??
0x55fb2c8877ee: testing::TestInfo::Run() @ ??:??
... Google Test internal frames ...

third_party/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:129
No Diff
Inline Diff
Line Diff
Token Diff
Char Diff
Ignore White Space
Value of: status.error_message()
Expected: has substring "Dimension 3 of size 3 is not evenly divisble by 4, at my_split"
  Actual: "Dimension 3 of size 3 is not evenly divisible by 4, at my_split"
Invalid argument: Dimension 3 of size 3 is not evenly divisible by 4, at my_split
Stack trace:
0x55fb2b36f5ad: tensorflow::tensorrt::convert::ExpectStatus(tensorflow::Status, tensorflow::error::Code, char const*) @ ../sysdeps/x86_64/start.S:121
0x55fb2b385435: tensorflow::tensorrt::convert::OpConverterTest::RunValidationAndConversion(tensorflow::NodeDef const&, tensorflow::error::Code, char const*, bool) @ ??:??
0x55fb2b3da8f0: tensorflow::tensorrt::convert::OpConverterTest_ConvertSplit_Test::TestBody() @ ??:??
0x55fb2c886521: testing::Test::Run() @ ??:??
0x55fb2c8877ee: testing::TestInfo::Run() @ ??:??
... Google Test internal frames ...

**Review by bixia1 — APPROVED** [review]
I fixed the test.

**Comment by DEKHTIARJonathan** [comment]
> I fixed the test.

Thanks Bixia

**Pull request merged**